### PR TITLE
Dl 1404

### PR DESCRIPTION
**Pull request by DINESHKUMARMURUGAN**

### CODE REVIEW

File: src/main/scala/com/gannett/lookalike/Segment.scala [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+import org.apache.spark.sql.{DataFrame, Dataset, Row, SparkSession}
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
+import org.apache.spark.transformers.{DataTransformer, LPConverter, NewSegmentedDataTransformer, SegmentedDataTransformer}
+import org.slf4j.{Logger, LoggerFactory}
+import hex.tree.gbm.{GBM, GBMModel}
+import hex.tree.gbm.GBMModel.GBMParameters
+import water.fvec.H2OFrame
+import water.{Key, MRTask}
+
+case class BroadcastVar() {
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/Segment.scala]: Why is this case class needed? What exactly is the use of this broadcast class?

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                                              ,VB.CLICK_NAME
+                                              ,VB.POST_PROP41
+                                              ,VB.PAGE_EVENT
+                                              ,VB.IS_CHILD
+                                              ,VB.EVENT_DATE
+                                              ,VB.USER_STATUS
+                                              ,COUNT (OMNITURE_ID) AS PAGE_VIEWS,
+                                              LOAD_DATE
+                                          FROM
+                                              OMNITURE_L1.BROWSER_VIEW_MCVISID VB
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: Is this view created in prod? If so how was it created, Is it updated every day?

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    val toLoaddate = try {
+      spark.sql("select DATE_ADD('"+toDate+"', 1)").first().get(0).toString
+    }
+    catch {
+      case e:Exception => toDate
+    }
+
+    val weekOfYear = spark.sql("select weekofyear('"+fromDate+"')").first().get(0).toString
+
+    val omni_l2_with_mcid = spark.sql("""SELECT
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: This dataframe doesnt seem to be used anywhere. Not even the temporary registered table.

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+      "search_engine_value, zip,product, page_views from omniture_l2.browser_page_view_counts " +
+      "where " +
+      "event_date between '"+fromDate+"' and '"+toDate+"' and load_date between '"+fromLoaddate+"' and '"+toLoaddate+"'")
+
+    omni_month.cache()
+    // omni_month.rdd.checkpoint()
+
+    omni_month.registerTempTable("month_content"+weekOfYear)
+
+    val omni_mon_summary = spark.sql("""select omniture_id, gup_anon_id, section, sub_section, topic,zip,
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: Instead of having 2 queries, on line 112 and this one, can this not be combined? Does spark optimize this?

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                                    count(distinct visit_num) as total_visit, sum(page_views) as total_pv
+                                    from month_content"""+weekOfYear+""" group by omniture_id, gup_anon_id,section, sub_section, topic,zip""")
+
+
+
+    //Remove Outliers
+    val omni_mon_summary_filtered = omni_mon_summary.filter(omni_mon_summary.col("total_pv") < 38)
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: Why 38? Where did this number come from? What if this changes?

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+
+    omni_mon_summary_filtered.registerTempTable("omni_mon_summary"+weekOfYear)
+
+    val omni_mon_explode = spark.sql("""select omniture_id,gup_anon_id,zip,
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: Will this query need to change if more ssts categories are added? Can there be a todo to convert this to work using a configuration of some sort? Like an improvement?

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+  def processDate(configuration: ProcessorConfiguration, spark: SparkSession,fromDate:String,toDate:String):Dataset[Row] = {
+
+      this.fromDate = fromDate
+      this.toDate = toDate
+    process(configuration,spark)
+  }
+  override def process(configuration: ProcessorConfiguration, spark: SparkSession): Dataset[Row] = {
+
+    val fromLoaddate = spark.sql("select DATE_SUB('"+fromDate+"', 1)").first().get(0).toString
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: Can we not use scala functions? Whats the reason of sql to generate dates?

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    val toLoaddate = try {
+      spark.sql("select DATE_ADD('"+toDate+"', 1)").first().get(0).toString
+    }
+    catch {
+      case e:Exception => toDate
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: If the toDate is passed to the processDate function, why is a new date created and assigned to toLoaddate? If a new date is to be used, why is toDate used in case of an exception?

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala [hunk 9]
```diff
@@ -0,0 +1,164 @@
+package com.gannett.lookalike.processors
+
+import com.gannett.lookalike.{DataProcessor, ProcessorConfiguration}
+import org.apache.spark.sql.{Dataset, Row, SparkSession}
+
+/**
+  * Created by dmurugan on 3/13/17.
+  */
+object OmnitureDataLoader extends Serializable with DataProcessor{
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: Documentation.

File: src/main/scala/com/gannett/lookalike/processors/DateProcessor.scala [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  * Created by dmurugan on 3/22/17.
+  */
+object DateProcessor extends Serializable with DataProcessor{
+  /**
+    * These date's are hardcoded since we use them as default for generating omniture data for the model
+    */
+  var fromDate = "2016-12-01"
+  var toDate ="2017-02-31"
+  def processForDate(forDate:String, configuration: ProcessorConfiguration, spark: SparkSession): Dataset[Row] = {
+    this.fromDate = spark.sql("select DATE_SUB('"+forDate+"', 90)").first().get(0).toString
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/DateProcessor.scala]: Is there an advantage of using sql for date manipulation?

File: src/main/scala/com/gannett/lookalike/processors/DateProcessor.scala [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+    this.toDate = forDate
+    process(configuration,spark)
+  }
+  override def process(configuration: ProcessorConfiguration, spark: SparkSession): Dataset[Row] = {
+
+    import spark.implicits._
+    var dataset:Dataset[Row] = null
+    var i = 1
+    while(fromDate < toDate) {
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/DateProcessor.scala]: Is there a reason to load omniture data in 7 day window and then calculating the average?

File: src/main/scala/com/gannett/lookalike/processors/aggregate/UserOmnitureDataLoader.scala [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+/**
+  * This class gets the three months aggregate of omniture data and adds it to the userdata
+  * This is a inner join - which means only the users from 3 months of omniture who has
+  * data in UserStore table is returned.
+  */
+object UserOmnitureDataLoader extends Serializable with DataProcessor{
+  override def process(configuration: ProcessorConfiguration, spark: SparkSession): Dataset[Row] = {
+
+    /**
+      * This one has to be updated - Currently it is fetching the data from the table - But it has to dynamically
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/aggregate/UserOmnitureDataLoader.scala]: Doesnt algo 0 do this?

**Comment by mpeshave** [comment]
Please add documentation on how to run this step by step in local as well as dev environment.

File: src/main/scala/com/gannett/lookalike/processors/aggregate/Group1UserDataLoader.scala [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    val omni_data = UserOmnitureDataLoader.process(configuration,spark)
+
+    val acxiom_data = AcxiomDataLoader.process(configuration,spark)
+
+    omni_data.registerTempTable("omniFinalTempTable")
+    acxiom_data.registerTempTable("acxiomFinalTempTable")
+
+
+    //Group 1 - For which all the values are present
+    val final360_data = spark.sql("""select a.omniture_id,a.emailaddress, a.zip, a.userguid, a.iseditorial, a.isactive, a.firstname, a.lastname,
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/aggregate/Group1UserDataLoader.scala]: Looks like Group1 user dataset is composed of 2 datasets,
a) UserOmnitureData - This data set is omniture browser data for the last 3 months joined with user store data.
b) Acxiom data - This data set is axciom data joined with: 1) Maax 2) Cirgds 3) Atypon 4) User-Omniture

Looks like UserStore-Omniture data is joined twice for generating group 1 data set. Why is user omniture dataset joined twice? Can the join not happen between UserOmniture and Acxiom data sets? This is slowing down the job?

File: src/main/scala/com/gannett/lookalike/Segment.scala [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+
+    spark.stop()
+  }
+
+
+
+
+
+
+  def testH2O(df:DataFrame,preddf:DataFrame) = {
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/Segment.scala]: Is this function just for testing H2O? What exactly does this function do? Some documentation and links to what H2O features are being used in this method would help the reader understand whats this function is for.

Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/Segment.scala]: Remove commented code.

Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/Segment.scala]: Instead of creating h2o context in the method and returning it back to the calling method to be stopped in the calling method, would it be better to create the h2o context in the calling method and send it as a param to this method? There could be failures if someone updated this method to stop h2o context.

File: src/main/scala/com/gannett/lookalike/Segment.scala [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+    val h2oContext = H2OContext.getOrCreate(spark.sparkContext, conf)
+
+
+    val hf: H2OFrame = h2oContext.asH2OFrame(df)
+
+    hf.colToEnum(Array("segment_id"))
+    val predhf: H2OFrame = h2oContext.asH2OFrame(preddf)
+
+    val keys = Array[String]("train.hex", "test.hex").map(Key.make[Frame](_))
+    val ratios = Array[Double](0.7, 0.3)
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/Segment.scala]: What is 0.7 and 0.3? Where did these numbers come from?

File: src/main/scala/com/gannett/lookalike/Segment.scala [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+    val frs = ShuffleSplitFrame.shuffleSplitFrame(hf, keys, ratios, 1234567689L)
```
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/Segment.scala]: What is 123456789L?

**Comment by mpeshave** [comment]
Remove unwanted models and code.

**Comment by mpeshave** [comment]
How long does this job take?

File: src/main/scala/com/gannett/lookalike/Segment.scala (hunk 1 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/Segment.scala]: Initially it was used to broacast the column names to all the executors - It is removed since the logic has been changed now.

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala (hunk 2 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: Not part of the logic - Since the Concept of MCVISID is temporarily removed. But the code is commented incase there is need for MCVISID in future

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala (hunk 3 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: Same as above

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala (hunk 4 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: I Just Kept it in separate to split the logic - since the logic for count kept changing. Spark takes the same time in both the cases(combined and separate). Another reason I kept them separate is to cache the data from omniture before grouping it.

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala (hunk 5 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: This is a part of exercise performed previously - It is 99th percentile of count per week. I will create a separate ticket for documentation and update it.

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala (hunk 6 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: As far as lookalike model goes - the change will happen only with respect to addition of known users in future. And there is no plan to change the no of columns for the model - which implies that we won't update the ssts. I'm also tagging @ishani28 for confirmation.

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala (hunk 7 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: Yup. Scala function can also be used. Just wrote it in sql to confirm things with Data science team. And since these are sql functions it can be tracked from spark UI - which will help us understand which part of a month is under process(Each SQL function that performs date is tracked as separate job)

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala (hunk 8 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: This boils down to the issue of load date and event date - In some cases where the event date data is on the next days load date. And for the exception I forgot the exact exception I was getting it was something with respect to last date of a month or a year.

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala (hunk 9 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: Added. For the overall documentation a Separate ticket will be created.

File: src/main/scala/com/gannett/lookalike/processors/DateProcessor.scala (hunk 11 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/processors/DateProcessor.scala]: Yes - Mainly to remove the outlier

File: src/main/scala/com/gannett/lookalike/processors/aggregate/UserOmnitureDataLoader.scala (hunk 12 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/processors/aggregate/UserOmnitureDataLoader.scala]: Yup - I'm updating the comment. Now Algo 0 will do

File: src/main/scala/com/gannett/lookalike/processors/aggregate/Group1UserDataLoader.scala (hunk 13 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/processors/aggregate/Group1UserDataLoader.scala]: Yes Useromniture and Acxiom is joined - in AtyponDataLoader object that feeds the data to Acxiom.

File: src/main/scala/com/gannett/lookalike/Segment.scala (hunk 14 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/Segment.scala]: Updating the name to runPredicion

Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/Segment.scala]: Removed

File: src/main/scala/com/gannett/lookalike/Segment.scala (hunk 16 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/Segment.scala]: Large Long Value - For random SEED value

File: src/main/scala/com/gannett/lookalike/Segment.scala (hunk 15 above)
Comment (DINESHKUMARMURUGAN) [inline src/main/scala/com/gannett/lookalike/Segment.scala]: This is a prediction technique which says take 70% as training set and 30% as test set. A separate ticket for documentation will be added.

File: src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala (hunk 3 above)
Comment (mpeshave) [inline src/main/scala/com/gannett/lookalike/processors/OmnitureDataLoader.scala]: Remove or comment un-used code.
